sdr_hackrf.c

```diff
@@ -0,0 +1,323 @@
+    HackRF.converter_state = NULL;
+}
+
+bool hackRFHandleOption(int argc, char **argv, int *jptr)
+{
+    int j = *jptr;
+    bool more = (j+1 < argc);
+    if (!strcmp(argv[j], "--lna-gain") && more) {
+        HackRF.lna_gain = atoi(argv[++j]);
+
+        if (HackRF.lna_gain % 8 != 0) {
+            fprintf(stderr, "Error: --lna-gain must be multiple of 8\n");
+            return false;
+        }
+
+        if (HackRF.lna_gain > 40 || HackRF.lna_gain < 0) {
+            fprintf(stderr, "Error: --lna-gain range is 0 - 42\n");
+            return false;
+        }
+    } else if (!strcmp(argv[j], "--vga-gain") && more) {
+        HackRF.vga_gain = atoi(argv[++j]);
+
+        if (HackRF.vga_gain % 2 != 0) {
+            fprintf(stderr, "Error: --vga-gain must be multiple of 2\n");
+            return false;
+        }
+
+        if (HackRF.vga_gain > 62 || HackRF.vga_gain < 0) {
+            fprintf(stderr, "Error: --vga-gain range is 0 - 62\n");
+            return false;
+        }
+
+    } else if (!strcmp(argv[j], "--ppm") && more) {
+        HackRF.ppm = atoi(argv[++j]);
+    } else if (!strcmp(argv[j], "--samplerate") && more) {
+        HackRF.rate = atoi(argv[++j]);
+    } else if (!strcmp(argv[j], "--enable-amp")) {
+        HackRF.enable_amp = 1;
+    } else {
+        return false;
+    }
+
+    *jptr = j;
+
+    return true;
+}
+
+void hackRFShowHelp()
+{
+    printf("      HackRF-specific options (use with --device-type hackrf)\n");
+    printf("\n");
+    printf("--enable-amp             enable amplifier)\n");
+    printf("--lna-gain               set LNA gain (Range 0-40 in 8dB steps))\n");
+    printf("--vga-gain               set VGA gain (Range 0-62 in 2dB steps))\n");
+    printf("--samplerate             set sample rate)\n");
+    printf("--ppm                    ppm correction)\n");
+    printf("\n");
+}
+
+static void show_config()
+{
+    fprintf(stderr, "freq : %ld\n", HackRF.freq);
+    fprintf(stderr, "lna_gain : %d\n", HackRF.lna_gain);
+    fprintf(stderr, "vga_gain : %d\n", HackRF.vga_gain);
+    fprintf(stderr, "samplerate : %d\n", HackRF.rate);
+    fprintf(stderr, "ppm : %d\n", HackRF.ppm);
+}
+
+bool hackRFOpen()
+{
+    if (HackRF.device) {
+        return true;
+    }
+
+    // Calculate sample rate and frequency deviation if ppm is specified
+    if (HackRF.ppm != 0) {
+        HackRF.rate = (uint32_t)((double)HackRF.rate * (1000000 - HackRF.ppm)/1000000+0.5);
+        HackRF.freq = HackRF.freq * (1000000 - HackRF.ppm)/1000000;
+    }
+
+    int status;
+
+    status = hackrf_init();
+    if (status != 0) {
+        fprintf(stderr, "HackRF: hackrf_init failed with code %d", status);
+        hackrf_close(HackRF.device);
+        hackrf_exit();
+        return false;
+    }
+
+    status = hackrf_open(&HackRF.device);
+    if (status != 0) {
+        fprintf(stderr, "HackRF: hackrf_open failed with code %d", status);
+        hackrf_close(HackRF.device);
+        hackrf_exit();
+        return false;
+    }
```

Comment:
If I understand the hackrf API correctly, the cleanup paths here need to be a little more careful (don't call exit if init failed; don't call close if open failed)